keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -0,0 +1,1018 @@
+run at night or over weekend without any human monitoring. In order to account
+for random infrastructure failures we define `.backoffLimit: 6` for the job.
+However, a signifficant portion of the failures happen due to bugs in code.
+Moreover, the failures may happen late during the program execution time. In
+such case, restarting such a pod results in wasting a lot of computational time.
+
+We would like to be able to automatically detect and terminate the jobs which are
+failing due to the bugs in code of the executable, so that the computation resources
+can be saved.
+
+#### Story 2
+
+As a service provider that offers computational resources to researchers I would like to
+have a mechanism which terminates jobs for which pods are failing due to user errors,
+but allows infinite retries for infrastructure errors (pod terminations by kubelet,
+kube-scheduler, abnormal node termination, etc).
+I do not have knowledge or influence over the executable that researchers run,
+so I don't know beforehand which exit codes they might return. This is important in order to limit the.
+
+<!--
+TODO: Include the API for both user stories once better defined
+-->
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!-- Increased complexity of the Job management, by interaction with some of
+already exiting Job configuration options. To mitigate this, we introduce
+minimal API which will allow to satisfy the main usage scenario.-->
+
+The Pod status (which includes exit codes) coule be lost if the failed pod is garbage collected.
+First, it should be rather rare so the overall consequence of unnecessary pod restarts
+will be limited. Second, we can prevent this by using the feature of
+[job tracking with finalizers](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers/).
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### JobSpec API
+
+```yaml
+apiVersion: v1
+kind: Job
+spec:
+  template:
+    spec:
+      containers:
+      - name: main-job-container
+        image: job-image
+        command: ["./program"]
+      - name: monitoring-job-container
+        image: job-monitoring
+        command: ["./monitoring"]
+  backoffLimit: 3
+  backoffPolicy:
+    onConditions:
+    - containerExitCodes: 1-12,100-127
+      containerName: main-job-container
+      action: Terminate
+    - podStatusReasons: ClusterScalingDown,JobPreemption
```

Comment:
clusterAutoscaler won't delete a pod. It would use the evictions API https://kubernetes.io/docs/concepts/scheduling-eviction/api-eviction/

Check what's the status reason in this case and use that.